STAR.py

```diff
@@ -1,29 +1,60 @@
+from ast import Starred
 import pandas as pd
 import numpy as np
-
+import copy
+from typing import Union
 class InvalidElection(Exception):
     pass
 
 class TrueTie:
     def __init__(self, tied):
         self.tied = tied
-        self.elected = []
-
-def summary_data(ballots: pd.DataFrame):
+        self.elected = [] 
+class SummaryData:
+    def __init__(self,score_hist, score_sums, pairwise_matrix, preference_matrix):
+        self.score_hist = score_hist
+        self.score_sums = score_sums
+        self.pairwise_matrix = pairwise_matrix
+        self.preference_matrix = preference_matrix
+
+    def drop(self,candidates):
+        self.score_hist.drop(candidates, axis=0, inplace=True)
+        self.score_sums.drop(candidates, axis=0, inplace=True)
+        self.pairwise_matrix.drop(candidates, axis=0, inplace=True)
+        self.pairwise_matrix.drop(candidates, axis=1, inplace=True)
+        self.preference_matrix.drop(candidates, axis=0, inplace=True)
+        self.preference_matrix.drop(candidates, axis=1, inplace=True)
+        return self
+
+    def keep(self,candidates):
+        self.score_hist = self.score_hist.loc[candidates, :]
+        self.score_sums = self.score_sums.loc[candidates]
+        self.pairwise_matrix = self.pairwise_matrix.loc[candidates, candidates]
+        self.preference_matrix = self.preference_matrix.loc[candidates, candidates]
+        return self
+
+def get_summary_data(ballots: pd.DataFrame):
     # The below quantities can be emitted however is most useful
-    score_hist = {a : ballots[a].value_counts() for a in ballots.columns}
+    score_hist = pd.DataFrame(0, columns = range(6), index = ballots.columns)
+    for a in ballots.columns:
+        score_hist.loc[a] = ballots[a].value_counts().sort_index()
+    score_hist = score_hist.fillna(0)
+    
     score_sums = ballots.sum()
-    pairwise_matrix = { a :
-            { b : pairwise_winner(ballots, a, b) for b in ballots.columns }
-        for a in ballots.columns
-    }
-    return (score_hist, score_sums, pairwise_matrix)
+    
+    preference_matrix = pd.DataFrame(0, columns = ballots.columns, index = ballots.columns)
+    for a in ballots.columns:
+        for b in ballots.columns:
+            preference_matrix.loc[a][b] = (ballots.loc[:][a]>ballots.loc[:][b]).sum()
+    pairwise_matrix = (preference_matrix > preference_matrix.T).astype(int)
```

Comment:
this measures "is strictly preferred," but should we add some way to denote ties?

Reply:
Ties could be determined by the transpose position. So if pairwise_matrix[a][b]==0 and pairwise_matrix[b][a]==0 then you have a tie. Or we could do the previous method of 1=win, 0=tie, -1=lose, would just take more than 1 line of code haha.

Reply:
I think it's ok actually; we handle the tied case anyway after line `146`